### PR TITLE
feat(api): return machine appropriate deck slot names in PAPIv2

### PR DESCRIPTION
**Pull request by jbleon95**
# Overview

Closes RLAB-337.

This PR makes it so that any call to a public facing Python method that previously would return a string for the deck slot (e.g. `"1"`, `"2"`, `"3"`, etc) now will return a deck coordinate (`"A1"`, `"A2"` `"A3"`, etc) when analyzing/running on a Flex machine. When on an OT-2, previous behavior will be kept.

In practice this only covers the two `parent` properties, `Labware.parent` and `ModuleContext.parent`. The `Deck` class also has methods that return slot definitions that at reference slot ids, but those are reliant on the deck definition, which is covered in PR #12571.

# Test Plan

Tested the following protocol passes analysis and runs on a robot

```
metadata = {
    'protocolName': 'Deck Coordinate PAPIv2 Test',
}

requirements = {
    "robotType": "OT-3",
    "apiLevel": "2.14"
}

LABWARE_LOCATIONS = ['2', 6, 'C1', 'b3']
OT3_COORDINATE_NAMES = ['D2', 'C3', 'C1', 'B3']

MOVE_LABWARE_LOCATION = 'C2'

THERMOCYCLER_LOCATION = 'B1'
HEATER_SHAKER_LOCATION = 'D3'
TEMP_MODULE_LOCATION_OT2 = 1
TEMP_MODULE_LOCATION_OT3 = 'D1'


def run(context):
    # Protocol Context Load Labware
    labwares = []
    for loc in LABWARE_LOCATIONS:
        labwares.append(context.load_labware("armadillo_96_wellplate_200ul_pcr_full_skirt", loc))

    # Labware Context (parent)
    for coordinate, labware in zip(OT3_COORDINATE_NAMES, labwares):
        assert labware.parent == coordinate, f"Labware parent {labware.parent} does not equal {coordinate}"

    # Protocol Context Move Labware
    context.move_labware(labware=labwares[0], new_location=MOVE_LABWARE_LOCATION, use_gripper=True)
    assert labwares[0].parent == MOVE_LABWARE_LOCATION, f"Labware parent {labwares[0].parent} does not equal {MOVE_LABWARE_LOCATION}"

    # Protocol Context Load Module
    thermocycler = context.load_module("thermocyclerModuleV2", THERMOCYCLER_LOCATION)
    heater_shaker = context.load_module("heaterShakerModuleV1", HEATER_SHAKER_LOCATION)
    temp_module = context.load_module("temperatureModuleV2", TEMP_MODULE_LOCATION_OT2)

    # Module Context (parent)
    assert thermocycler.parent == THERMOCYCLER_LOCATION, f"Thermocycler parent does not equal {THERMOCYCLER_LOCATION}"
    assert heater_shaker.parent == HEATER_SHAKER_LOCATION, f"Heater-Shaker parent does not equal {HEATER_SHAKER_LOCATION}"
    assert temp_module.parent == TEMP_MODULE_LOCATION_OT3, f"Temperature Module parent does not equal {TEMP_MODULE_LOCATION_OT3}"
```

# Changelog

- Updated `Labware.parent` and `ModuleContext.parent` to return deck coordinates when running on Flex
- Removed some implicit `DeckSlotName` to `str` conversions that were unclear
- Added `robot_type` property to protocol cores
- Added OT-3 integration test to robot server for analyzing a protocol returning deck coordinates
- Fixed workflow oversight where tests marked as `ot3_only` were never getting run on CI

# Review requests

# Risk assessment

Medium. This PR does not change behavior on the OT-2, but any protocol that relies on any of the above methods may not function as they did previous to this PR.

### CODE REVIEW

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 1]
```diff
@@ -214,7 +214,7 @@ def labware(self) -> Optional[Labware]:
     @requires_version(2, 14)
     def parent(self) -> str:
         """The name of the slot the module is on."""
-        return self._core.get_deck_slot().value
+        return self._core.get_deck_slot_display_name()
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/module_contexts.py]: Hmm. Returning something called the "display name" from `.parent` makes my spidey senses tingle.

`ModuleContext.parent` seems like the kind of thing that needs to return strict and machine-readable values, because people can build logic around it. Meanwhile, "display name," at least in deck definitions, is not that.

https://github.com/Opentrons/opentrons/blob/1484ce035f72dc9441803f85d9aba66f704dc3ca/shared-data/deck/schemas/3.json#L157-L160

https://github.com/Opentrons/opentrons/blob/1484ce035f72dc9441803f85d9aba66f704dc3ca/shared-data/deck/definitions/3/ot3_standard.json#L24

In other words, I think `.parent` *should* return a string like `"A1"`, and I think it technically *does* in this PR, but the implementation *looks like* it's going to return a string like `"Slot A1"`.

Comment (SyntaxColoring) [reply api/src/opentrons/protocol_api/module_contexts.py]: Am I just getting tripped up by the naming of the `get_deck_slot_display_name()` function? Would it be correct to rename it `get_deck_slot_id()`, referring to the ID as it appears in the deck definition?

File: api/src/opentrons/protocol_api/core/legacy/legacy_protocol_core.py [hunk 2]
```diff
@@ -90,6 +91,10 @@ def api_version(self) -> APIVersion:
         """Get the API version the protocol is adhering to."""
         return self._api_version
 
+    @property
+    def robot_type(self) -> RobotType:
+        return "OT-2 Standard"
+
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/core/legacy/legacy_protocol_core.py:97]: 👍 Yeah this seems reasonable.

File: robot-server/tests/integration/conftest.py [hunk 3]
```diff
@@ -91,10 +91,24 @@ def set_disable_fast_analysis(
...
+    request_session: requests.Session,
+) -> Iterator[None]:
+    """For integration tests that need to set then clear the
+    enableOT3HardwareController feature flag"""
+    url = "http://localhost:31950/settings"
+    data = {"id": "enableOT3HardwareController", "value": True}
+    request_session.post(url, json=data)
+    yield None
+    data["value"] = None
+    request_session.post(url, json=data)
```
Comment (SyntaxColoring) [inline robot-server/tests/integration/conftest.py]: I think implementing this by `POST`ing a feature flag is going to be fragile.

We're pretty certain that options that specify "what kind of robot is this server running on" should really be *server launch options,* not runtime-mutable feature flags. (See https://github.com/Opentrons/opentrons/pull/11812#discussion_r1054540176, for example.) We should get into the habit of doing that now so we don't make things unnecessarily difficult and confusing for us to fix in the future.

I also think this will lead to inconsistent internal server state. I guess this does successfully change the configuration of `ProtocolEngine`s used in subsequent analyses. But it *won't* affect the server's global singleton `HardwareControlAPI`, which is used by runs. That will still be initialized as an OT-2 during server startup, and it will remain that for the lifetime of the server.

I think for this PR, we should:

1. Update [`DevServer`](https://github.com/Opentrons/opentrons/blob/1484ce035f72dc9441803f85d9aba66f704dc3ca/robot-server/tests/integration/dev_server.py#L11) to take an `is_ot3` argument or something. If `True`, it sets `OT_API_FF_enableOT3HardwareController` when launching the process.
2. Make an `ot3_run_server` session-scoped fixture to match `run_server`. It will have to run on some other arbitrary port.
3. Make an `ot3_clean_server_state` fixture to match `clean_server_state`.

And then, some time later:

4. Whenever we get around to it, merge `clean_server_state` with `run_server` and merge `clean_ot3_server_state` with `run_ot3_server`, for ergonomics.
5. When there's a proper robot-server launch option to set the robot type, we can easily update `DevServer` to use that instead of the feature flag environment variable.

**Review by SyntaxColoring — COMMENTED** [review]
Lookin' good so far!

File: robot-server/tests/integration/conftest.py [hunk 4]
```diff
@@ -77,6 +78,40 @@ def run_server(
...
+) -> Iterator[None]:
+    """Run the robot server in a background process."""
+    with DevServer(
+        port=_OT3_SESSION_SERVER_PORT,
+        is_ot3=True,
+        ot_api_config_dir=Path(server_temp_directory),
+    ) as dev_server:
+        dev_server.start()
+
+        # Wait for a bit to get started by polling /hcpealth
```
Comment (SyntaxColoring) [inline robot-server/tests/integration/conftest.py:94]: Sorry this is such a duplicated mess at the moment, but would you mind merging the changes that https://github.com/Opentrons/opentrons/pull/12638/files made to `run_server` into this new `ot3_run_server` fixture?

File: robot-server/tests/integration/http_api/protocols/test_deck_coordinate_load.tavern.yaml
Comment (SyntaxColoring) [inline robot-server/tests/integration/http_api/protocols/test_deck_coordinate_load.tavern.yaml:1]: Recapping from an in-person between me and @jbleon95, for posterity:

We're basically integration-testing the Python Protocol API through robot-server here. We're doing this because I don't know what the proper way to do this is, these days. But this is almost certainly not it. We should keep thinking about how to do this better.

File: api/tests/opentrons/protocol_api/test_labware.py [hunk 5]
```diff
@@ -194,7 +186,7 @@ def test_parent_module_context(
         mock_temp_module_context
     )
 
-    subject.parent == mock_temp_module_context
+    assert subject.parent == mock_temp_module_context
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_api/test_labware.py:189]: 😬 Nice catch.

File: api/src/opentrons/protocol_api/core/legacy/legacy_module_core.py [hunk 6]
```diff
@@ -87,6 +87,9 @@ def get_deck_slot(self) -> DeckSlotName:
         """Get the module's deck slot."""
         return DeckSlotName.from_primitive(self._geometry.parent)  # type: ignore[arg-type]
 
+    def get_deck_slot_id(self) -> str:
+        return str(self.get_deck_slot())
+
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/core/legacy/legacy_module_core.py]: Nitpicky, sorry:

```suggestion
    def get_deck_slot_id(self) -> str:
        return self.get_deck_slot().id
```

File: .github/workflows/robot-server-lint-test.yaml [hunk 7]
```diff
@@ -76,8 +76,12 @@ jobs:
         run: make -C robot-server setup-ot2
       - name: Lint
         run: make -C robot-server lint
-      - name: Test
+      - if: ${{ matrix.with-ot-hardware == 'false' }}
+        name: Test without opentrons_hardware
         run: make -C robot-server test-cov test_opts="-m 'not ot3_only'"
+      - if: ${{ matrix.with-ot-hardware == 'true' }}
+        name: Test with opentrons_hardware
+        run: make -C robot-server test-cov test_opts="-m 'not ot2_only'"
```
Comment (SyntaxColoring) [inline .github/workflows/robot-server-lint-test.yaml]: Nice catch.

The original addition of `not ot2_only` feels weird to me. Why are we excluding entire server tests when `opentrons_hardware` is missing, anyway? If anything, shouldn't we test that the endpoint returns an explicit "Not an OT-3" error, or something?

Anyway. For this PR, when we're testing with `opentrons_hardware`, I think we should just:

```suggestion
      - if: ${{ matrix.with-ot-hardware == 'true' }}
        name: Test with opentrons_hardware
        run: make -C robot-server test-cov
```

**Review by SyntaxColoring — APPROVED** [review]
Nice. Thanks for figuring out how to split this into an individually mergeable piece. Here are a couple of small things, but other than these, this looks good to merge.
